scvi/external/cellassign/_module.py

```diff
@@ -0,0 +1,188 @@
+        mu_ngc = torch.exp(log_mu_ngc)  # (n, g, c)
+
+        # compute basis means for phi - shape (B)
+        basis_means_fixed = np.linspace(torch.min(x), torch.max(x), self.B)
+        basis_means = torch.tensor(basis_means_fixed)  # (B)
+
+        # compute phi of NegBin - shape (n_cells, n_genes, n_labels)
+        a = torch.exp(self.log_a)  # (B)
+        a_e = a.expand(s.shape[0], self.n_genes, self.n_labels, self.B)
+        b_init = 2 * ((basis_means_fixed[1] - basis_means_fixed[0]) ** 2)
+        b = torch.exp(torch.ones(self.B) * (-np.log(b_init)))  # (B)
+        b_e = b.expand(s.shape[0], self.n_genes, self.n_labels, self.B)
+        mu_ngc_u = mu_ngc.unsqueeze(-1)
+        mu_ngcb = mu_ngc_u.expand(
+            s.shape[0], self.n_genes, self.n_labels, self.B
+        )  # (n, g, c, B)
+        basis_means_e = basis_means.expand(
+            s.shape[0], self.n_genes, self.n_labels, self.B
+        )  # (n, g, c, B)
+        phi = (  # (n, g, c)
+            torch.sum(a_e * torch.exp(-b_e * torch.square(mu_ngcb - basis_means_e)), 3)
+            + self.LOWER_BOUND
+        )
+
+        # compute gamma
+        nb_pdf = NegativeBinomial(probs=mu_ngc, total_count=phi)
+        y_ = x.unsqueeze(-1).expand(s.shape[0], self.n_genes, self.n_labels)
+        y_log_prob_raw = nb_pdf.log_prob(y_)  # (n, g, c)
+        theta_log_e = theta_log.expand(s.shape[0], self.n_labels)
+        p_y_unorm = torch.sum(y_log_prob_raw, 1) + theta_log_e  # (n, c)
+        p_y_norm = torch.logsumexp(p_y_unorm, 1)
+        p_y_norm_e = p_y_norm.unsqueeze(-1).expand(s.shape[0], self.n_labels)
+        gamma = torch.exp(p_y_unorm - p_y_norm_e)  # (n, c)
+
+        return dict(
+            mu=mu_ngc,
+            phi=phi,
+            gamma=gamma,
+            p_y_unorm=p_y_unorm,
+            s=s,
+        )
+
+    def loss(
+        self,
+        tensors,
+        inference_outputs,
+        generative_outputs,
+        n_obs: int = 1.0,
+    ):
+        # generative_outputs is a dict of the return value from `generative(...)`
+        # assume that `n_obs` is the number of training data points
+        p_y_unorm = generative_outputs["p_y_unorm"]
+        gamma = generative_outputs["gamma"]
+
+        # compute Q
+        # take mean of number of cells and multiply by n_obs (instead of summing n)
+        q_per_cell = torch.sum(gamma * p_y_unorm, 1)
+
+        # third term is log prob of prior terms in Q
+        theta_log = F.log_softmax(self.theta_logit)
+        theta_log_prior = Dirichlet(torch.tensor(self.dirichlet_concentration))
+        theta_log_prob = -theta_log_prior.log_prob(
+            torch.exp(theta_log) + self.THETA_LOWER_BOUND
+        )
+        delta_log_prior = Normal(self.delta_log_mean, self.delta_log_variance)
+        summed_delta_log = torch.sum(self.delta_log)
+        delta_log_prob = -torch.sum(delta_log_prior.log_prob(summed_delta_log))
+        prior_log_prob = theta_log_prob
+        if self.shrinkage:
```

Comment:
is there a logical issue if `self.shrinkage` is False, then `delta_log_prior = Normal(self.delta_log_mean, self.delta_log_variance)` would give a runtimeerror